cpp/src/arrow/util/hashing.h

```diff
@@ -691,7 +691,7 @@ class BinaryMemoTable : public MemoTable {
 
     const builder_offset_type* offsets = binary_builder_.offsets_data();
     const builder_offset_type delta =
-        start < binary_builder_.value_data_length() ? offsets[start] : 0;
+        start < binary_builder_.length() ? offsets[start] : 0;
```

Comment:
Perhaps the proper condition should be `binary_builder_.length() != 0 ? offsets[start] : 0`?

Reply:
`binary_builder_.length() != 0` is the minimal case we need to fix, and we still do out-of-bound access (and write the data read to `out_data`) when `start >= binary_builder_.length()`.
If we do `start < binary_builder_.length()`, we cover the case when `start >= binary_builder_.length()` as well.

Reply:
Is `start >= binary_builder_.length()` possible though? And in that case, should `delta` really be 0?

Reply:
In case of `binary_builder_.length() > 0`, caller is responsible to make sure `start < binary_builder_.length()`.  So it's ok to just check `binary_builder_.length() > 0`.  It's just a little bit safer to check `start < binary_builder_.length()` ourselves.

Reply:
I'd rather use the strict check. It's not obvious that `delta = 0` is the right choice otherwise (since the table is non-empty).

Reply:
Depends on if you want to allow `start >= binary_builder_.length()`.  If it is allowed and the range is selected as ![formula](https://render.githubusercontent.com/render/math?math=index%20\in%20(-\infty,%20length)%20\cap%20[start,%20%2b\infty)), it is ok to set `delta = 0`, and leave the only element in `out_data` as `binary_builder_.value_data_length()`.  If we set `delta = offsets[start]` in this case, we are going to copy some unauthorized memory to `out_data` and it's a potential vulnerability.

Reply:
Good point, then I agree the current fix is fine. Thank you for the explanation!